Refactor parallel_scan() tests using RangePolicy

tcclevenger opened the pull request:
Continuation of Issue https://github.com/kokkos/kokkos/issues/5560.

Gather all tests of `parallel_scan()` using `RangePolicy` as the `ExecutionPolicy` into one test. This reduces redundancy that existed in `TestScan.hpp`, `TestRange.hpp`, and `TestRangeRequire.hpp`. 

I chose to eliminate `TestScan.hpp` and create `TestRangePolicyScan.hpp` in its place, since the former tested `parallel_scan()` only when `RangePolicy` was used. My plan is to have a future PR which creates/organizes `TestTeamPolicyScan` and `TestTaskSchedulerScan` which complete the tests of `parallel_scan()` functions.

Of course, please give any feedback and disagreement over what I have here.

Rombur commented on core/unit_test/CMakeLists.txt:
```diff
@@ -359,25 +359,26 @@ if(Kokkos_ENABLE_OPENACC)
+    ${CMAKE_CURRENT_BINARY_DIR}/openacc/TestOpenACC_MDRangePolicyConstructors.cpp
     ${CMAKE_CURRENT_BINARY_DIR}/openacc/TestOpenACC_Other.cpp
+    ${CMAKE_CURRENT_BINARY_DIR}/openacc/TestOpenACC_RangePolicyConstructors.cpp
+    ${CMAKE_CURRENT_BINARY_DIR}/openacc/TestOpenACC_RangePolicyScan.cpp
```
Are you sure `RangePolicyScan` needs to be removed? `Scan` was not in the list.

Rombur commented on core/unit_test/TestRangePolicyScan.hpp:
```diff
@@ -0,0 +1,227 @@
+#ifndef KOKKOS_ENABLE_OPENMPTARGET
+TEST(TEST_CATEGORY, range_policy_scan) {
+  {
+    TestRangePolicyScan<char> f;
```
Can we add `TestRangePolicyScan<bool>`.  This would have helped catch a bug in the HIP backend

Rombur replied:
Actually we shouldn't do that. It makes no sense to do a parallel_scan on boolean. I was thinking about parallel_reduce when I wrote that. Sorry for the confusion

Rombur commented:
@tcclevenger I see from the latest commit that `openacc` does not work with `bool`. Instead of removing the entire test, add a `FIXME_OPENACC` and guard the failing test with `#ifndef KOKKOS_ENABLE_OPENACC`.

cc: @seyonglee @pedrovalerolara

tcclevenger commented:
@Rombur Thanks for the comments. I'm checking if `#ifndef KOKKOS_ENABLE_OPENMPTARGET` is also needed. It was used in some of the previous tests, but not others. Even though all tests were using `RangePolicy`. Maybe there is an issue when setting a `Kokkos::Schedule<>`? I'm not exactly sure what triggers the issues with OPENMP_TARGET yet.

Rombur commented:
Can we just drop the last four commits, so everybody will forget the mess we made  :D It's also easier when we do a `git bisect` if the tests are passing in more commits.

tcclevenger commented:
@Rombur Yes, that is a good idea. I think the "remove RangePolicyScan from openacc remove list" is still valid. And I also _think_ I do not need the ifdef guard for openmptarget. I'll clean up the commit history.

Rombur commented:
Great thanks

tcclevenger commented:
Converting this to draft while I figure out the continuous-integration/jenkin-ci/pr failures

tcclevenger commented:
Only one test instance is failing, and I do not believe it is related to this PR. `OPENACC-NVHPC-CUDA` fails with
```
NVC++-S-1061-Procedures called in a compute region must have acc routine information - erf (/var/jenkins/workspace/Kokkos/build/core/unit_test/openacc/TestOpenACC_MathematicalFunctions2.cpp: 438)
NVC++/x86-64 Linux 22.3-0: compilation completed with severe errors
make[2]: *** [core/unit_test/CMakeFiles/KokkosCore_UnitTest_OpenACC.dir/build.make:244: core/unit_test/CMakeFiles/KokkosCore_UnitTest_OpenACC.dir/openacc/TestOpenACC_MathematicalFunctions2.cpp.o] Error 2
make[2]: *** Waiting for unfinished jobs....
```
This could be fixed by excluding the `TestOpenACC_MathematicalFunctions2` in CMakeLists, if this is a recurring issue on many PRs.

masterleinad commented on core/unit_test/CMakeLists.txt:
```diff
@@ -428,17 +415,32 @@ IF(KOKKOS_ENABLE_OPENMPTARGET AND KOKKOS_CXX_COMPILER_ID STREQUAL NVHPC)
+    ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_double_reduce_dynamic.cpp
+    ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_HostSharedPtr.cpp
+    ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_HostSharedPtrAccessOnDevice.cpp
+    ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_int64_t_reduce.cpp
+    ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_int64_t_reduce_dynamic.cpp
+    ${CMAKE_CURRENT_BINARY_DIR}/openmptarget/TestOpenMPTarget_int64_t_reduce_dynamic_view.cpp
```
Do these even exist?

tcclevenger replied:
Looks like these are the names of tests, e.g., `TEST(TEST_CATEGORY, int64_t_reduce)`. I'll remove them.

masterleinad reviewed (changes requested):
It appears that you accidentally enabled deleted
- ${CMAKE_CURRENT_BINARY_DIR}/openacc/TestOpenACC_MathematicalFunctions1.cpp
- ${CMAKE_CURRENT_BINARY_DIR}/openacc/TestOpenACC_MathematicalFunctions2.cpp

enabling these tests which make the CI fail.

dalg24 commented on core/unit_test/TestRangePolicyScan.hpp:
```diff
@@ -0,0 +1,228 @@
+namespace {
+
+template <typename ValueType>
+struct TestRangePolicyScan {
```
Per our discussion on Slack, I think `TestParallelScanRangePolicy` is more intuitive.  Also applies to file names.

dalg24 commented on core/unit_test/TestParallelScanRangePolicy.hpp:
```diff
@@ -0,0 +1,228 @@
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+#include <cstdio>
```
Make sure you include the header `<gtest/gtest.h>` explicitly here

tcclevenger commented:
Retest this please.

tcclevenger commented:
Retest this please

dalg24 commented:
Retest this please

dalg24 commented on core/unit_test/CMakeLists.txt:
```diff
@@ -113,7 +113,7 @@ foreach(Tag Threads;Serial;OpenMP;Cuda;HPX;OpenMPTarget;OpenACC;HIP;SYCL)
     set(${Tag}_SOURCES1A)
     foreach(Name
-        BitManipulationBuiltins
+        Abort
```
This reordering in alphabetical order makes me a bit nervous.  Please submit it as a separate PR

tcclevenger replied:
Yep, I've already run issue an issue before. I'll create a new one and let that merge first.

tcclevenger commented:
New PR https://github.com/kokkos/kokkos/pull/5931 takes care of the reorg of the CMakelists.txt. I rebased this PR on that branch and cleaned up the history/commit names. I'll leave this as draft until that one is merged.

Rombur commented on core/unit_test/Makefile:
```diff
@@ -127,7 +127,7 @@ ifeq ($(KOKKOS_INTERNAL_USE_CUDA), 1)
     OBJ_CUDA += TestCuda_SubView_c10.o TestCuda_SubView_c11.o TestCuda_SubView_c12.o
     OBJ_CUDA += TestCuda_SubView_c13.o
-    OBJ_CUDA += TestCuda_Reductions.o TestCuda_Scan.o
+    OBJ_CUDA += TestCuda_Reductions.o TestCuda_ParallelScanRangePolicy.o
```
I see all the backend but HIP.  It looks like we didn't have `TestHIP_Scan` but that's a mistake

masterleinad reviewed (approved):
Looks OK to me.
